protocols/autonat/src/behaviour.rs

```diff
@@ -478,35 +479,74 @@ impl NetworkBehaviour for Behaviour {
         self.inner.addresses_of_peer(peer)
     }
 
-    fn inject_event(
-        &mut self,
-        peer_id: PeerId,
-        conn: ConnectionId,
-        event: RequestResponseHandlerEvent<AutoNatCodec>,
-    ) {
-        self.inner.inject_event(peer_id, conn, event)
-    }
-
-    fn inject_listen_failure(
-        &mut self,
-        local_addr: &Multiaddr,
-        send_back_addr: &Multiaddr,
-        handler: Self::ConnectionHandler,
-    ) {
-        self.inner
-            .inject_listen_failure(local_addr, send_back_addr, handler)
-    }
-
-    fn inject_new_listener(&mut self, id: ListenerId) {
-        self.inner.inject_new_listener(id)
-    }
-
-    fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
-        self.inner.inject_listener_error(id, err)
-    }
-
-    fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &std::io::Error>) {
-        self.inner.inject_listener_closed(id, reason)
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::ConnectionEstablished {
+                peer_id,
+                connection_id,
+                endpoint,
+                failed_addresses,
+                other_established,
```

Comment:
We might want to consider introduced structs for each enum variant so we can just pass along the struct here instead of each field.

For example:

```rust
enum FromSwarm {
	ConnectionEstablished(ConnectionEstablished)
}

struct ConnectionEstablished {
	peer_id: PeerId,
	// etc
}
```

Thoughts?

Reply:
yeah, makes sense to me also, wdyt @mxinden?

Reply:
Works for me. If I am not mistaken one can pattern match these within the `match` arms and thus there is no downside for users that don't delegate to specific methods, right? Something along the lines of:

``` rust
match event {
  FromSwarm::ConnectionEstablished(ConnectionEstablished { peer_id, .. }) => { todo!() },
```

Reply:
Yep that is possible but a bit more verbose.

Reply:
but if we are using structs we should define getters and setters to not expose its inner fields right? and therefore struct fields will not be able to be destructured on `match`. Would that be a problem @mxinden?

Reply:
> but if we are using structs we should define getters and setters to not expose its inner fields right?

Not necessarily. Making fields of a struct private is useful if e.g. there are invariants we need to maintain in how they can be modified. In our case, these structs are just "bags" of data so the fields can easily be public IMO. Really what we would need is [enum-variant types](https://github.com/rust-lang/lang-team/issues/122).